sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala

```diff
@@ -72,6 +72,47 @@ object DatasetBenchmark {
     benchmark
   }
 
+  def backToBackMapPrimitive(spark: SparkSession, numRows: Long, numChains: Int): Benchmark = {
+    import spark.implicits._
+
+    val df = spark.range(1, numRows).select($"id".as("l"))
+    val benchmark = new Benchmark("back-to-back map for primitive", numRows)
+    val func = (d: Long) => d+1
+
+    val rdd = spark.sparkContext.range(1, numRows).map(l => l.toLong)
+    benchmark.addCase("RDD") { iter =>
+      var res = rdd
+      var i = 0
+      while (i < numChains) {
+        res = rdd.map(func)
+        i += 1
+      }
+      res.foreach(_ => Unit)
+    }
+
+    benchmark.addCase("DataFrame") { iter =>
+      var res = df
+      var i = 0
+      while (i < numChains) {
+        res = res.select($"l" + 1 as "l")
+        i += 1
+      }
+      res.queryExecution.toRdd.foreach(_ => Unit)
```

Comment:
this is what i'm talking about -- this forces a conversion, which is very expensive.